add option to change GENERATION_SUFFIX

dodgex opened the pull request:
see https://github.com/excilys/androidannotations/issues/1209

this allows to set the `generationSuffix` via options. at the moment the same valule is used for all generated stuff (classes, fields, methods) where it before was `_`. But for a first look i already opened the PR.

WonderCsabo commented on AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/helper/ModelConstants.java:
```diff
@@ -49,4 +51,12 @@
 	}
 
+	public static void init(ProcessingEnvironment processingEnv) {
+		OptionsHelper optionsHelper = new OptionsHelper(processingEnv);
```
@dodgex you can use `SourceVersion.isName` method to the the validation.

dodgex replied:
and what SourceVersion value to use? `SourceVersion.RELEASE_6` or `SourceVersion.RELEASE_7`?

WonderCsabo replied:
https://github.com/excilys/androidannotations/blob/326fe9666df37284245c4e2706f4c042f0b396c1/AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/AndroidAnnotationProcessor.java#L285

WonderCsabo replied:
BTW, isName is static, isn't it? The enum value does not matter.

WonderCsabo replied:
So `if (SourceVersion.isName(classSuffix) || classSuffix.contains("."))`

dodgex replied:
ah. missed that static... well i'll update the PR when i get the time.

dodgex replied:
updated

dodgex commented on AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/helper/ValidatorHelper.java:
```diff
@@ -26,7 +26,6 @@
 import static org.androidannotations.helper.CanonicalNameConstants.HTTP_MESSAGE_CONVERTER;
 import static org.androidannotations.helper.CanonicalNameConstants.INTERNET_PERMISSION;
 import static org.androidannotations.helper.CanonicalNameConstants.WAKELOCK_PERMISSION;
-import static org.androidannotations.helper.ModelConstants.GENERATION_SUFFIX;
```
the static import is there. but some lines below.

WonderCsabo commented:
This needs and update due to #1325.

dodgex commented:
it needs. `generationSuffix()` is for everything that is not a class.

WonderCsabo commented:
Then this PR is totally incomplete, am i right? I just quickly searched in Eclipse "*_*", and it returned lots of results which are not included here.

dodgex commented:
well i have no idea anymore. but yeah. for some reason are some `_` missing. but if i remember correctly there are some that do not need to be changed as they have a small scope (like a local variable inside an method).

i'll try to get to this one the next days and update it if i can get some time.

dodgex commented:
updated the PR. but i only changed thos "_" that are not inside a small block like a method. so all generated methods and fields should now use the `generationSuffix()` but locals and params not. not sure if that is needed?

dodgex commented:
hmf something is wrong with the formatter/save acitons in the eclipse. crtl-shift-f formats the source correctly but saving removes the indentation.

from what i found it seems to be the `additional actions -> correct indentation` option. @WonderCsabo can you check if it does that for you too?  (i tested in EFragmentHolder.java#130)

WonderCsabo commented:
Saving removes some indentation for me. Notably in case of chained methods.

dodgex commented:
any chance to get this merged, or is there still something to change? rebasing this PR is not fun. :D

WonderCsabo commented:
Thanks for your work on this, and sorry for the lots of rebases. 

> updated the PR. but i only changed thos "_" that are not inside a small block like a method. so all generated methods and fields should now use the generationSuffix() but locals and params not. not sure if that is needed?

Two things:

1. Maybe we could just change anything (locals and params, too). But i am not sure. @yDelouis?
2. I think you are missing fields and methods, too. I just searched for ```_"``` in Eclipse (no regex). And i found lots of results which are fields and methods.

dodgex commented:
wtf... updated the PR with those that where missing. but i'm pretty sure that i had changed all previously. maybe rebase broke it. would not be the first time.

WonderCsabo commented:
It seems all needed things changed now. Did you test this PR? I mean changing the suffix and see the functional test still green? This would need the rewrite of the whole functional test suite though. :(

dodgex commented:
changing the suffix can be considered as a breaking change. the tests are only testing the default suffix. but all these test pass when i run `mvn clean verify`

WonderCsabo commented:
I mean somehow we should verify that changing the suffix is working! Can you at least add a compile time test? Add a processor parameter which changes the suffix, then verify the new class is generated.

dodgex commented:
working on a test. currenlty we have `classSuffix` used for class names and `generationSuffix` for fields and methods.

in the current implementation only `classSuffix` is changable with a parameter. should i add an option for `generationSuffix` and test it or do we only provide the ability to change `classSuffix` as the user should never get in touch with anything that uses `generationSuffix` as it is only used for internal methods & fields?

WonderCsabo commented:
`generationSuffix` should be never changed, the purpose of it is to make the suffix consistent through the codebase. It is enough to test `classSuffix` change. Refactoring `generationSuffix` is already tested in the functional test project and it passed. :)

dodgex commented:
test added

WonderCsabo commented:
@dodgex i am wondering, should'nt we remove the `_` suffices from the locals and params?

WonderCsabo commented on AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/EFragmentHolder.java:
```diff
@@ -134,7 +136,7 @@ private void setFindViewById() {
 
 	private void setFragmentBuilder() throws JClassAlreadyExistsException {
-		fragmentBuilderClass = generatedClass._class(PUBLIC | STATIC, "FragmentBuilder_");
+		fragmentBuilderClass = generatedClass._class(PUBLIC | STATIC, "FragmentBuilder" + classSuffix());
```
I do not think it should use `classSuffx`, but `generationSuffix` instead.

dodgex replied:
why? its an generated class that is accessible for the user. so `FragmentBuilderImpl` is better than `FragmentBuilder_` actually i'd rename the [`FragmentBuilder`](https://github.com/excilys/androidannotations/blob/326fe9666df37284245c4e2706f4c042f0b396c1/AndroidAnnotations/androidannotations-api/src/main/java/org/androidannotations/api/builder/FragmentBuilder.java) api class (to e.g. `AbstractFragmentBuilder`) and call the generated class FragmentBuilder so this would be CLEAN.

same to other Builder classes.

WonderCsabo replied:
I do not agree. Generally we use this impl classes when there is an interface for a specific work, and then we provide an implementation for that work. Here, we do not provide implementation for the work, but we generate always the interface actually, too.

dodgex replied:
ok. then i'll change that.

WonderCsabo commented on AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/EViewHolder.java:
```diff
@@ -155,6 +156,6 @@ public JFieldVar getAlreadyInflated() {
 
 	private void setAlreadyInflated() {
-		alreadyInflated = generatedClass.field(PRIVATE, JType.parse(codeModel(), "boolean"), "alreadyInflated_", JExpr.FALSE);
+		alreadyInflated = generatedClass.field(PRIVATE, JType.parse(codeModel(), "boolean"), "alreadyInflated" + generationSuffix(), JExpr.FALSE);
```
You should also used this in the JavaDoc which is added because of this field. You can leave that constant, and replace `_` with `generationSuffix()`.

dodgex replied:
i hope i understood you correctly. :)

WonderCsabo replied:
:+1:

WonderCsabo commented:
@dodgex any update on this?

dodgex commented:
@WonderCsabo i'm currently not at home. Will try to update this PR next week.

WonderCsabo commented:
Thanks for your work on this! Can you update the wiki?

WonderCsabo commented:
Thanks to @dodgex for editing the wiki.

chpasha commented:
>i do not think we should remove the suffices from anywhere. The suffix also clearly shows that the class/etc. is generated by us, which is very important if you have a bigger project and/or multiple people are working on it

Hi guys, regarding "clearly showing that the class is generated by us" - why not use @Generated annotation which was meant exactly for that? http://docs.oracle.com/javaee/5/api/javax/annotation/Generated.html . As far as I understand, it is enough to modify setGeneratedClass-Method inside BaseGeneratedClassHolder adding a line like this 
>generatedClass.annotate(Generated.class)
				.param("value", AndroidAnnotationProcessor.class.getName())
                                .param("date", getISO8601StringForDate(new Date()));

maybe even use a specific holder FQN instead of generic AndroidAnnotationProcessor.class

WonderCsabo commented:
@chpasha that annotation is not available in the Android runtime, and will cause problems if we add it, despite its source only retention. BTW lombok has a `lombok.addGeneratedAnnotation` property for this reason.

chpasha commented:
I didn't know that, so I just was hacking around and got no errors when running some test android project with this annotation applied. Which problems will it cause since it is discarded by compiler?

WonderCsabo commented:
Okay, it seems `@Generated` is not a problem, i mistaken it for `@ConstructorProperties`.

In that case, i agree with you @chpasha.

chpasha commented:
great :)

WonderCsabo commented:
@chpasha unfortunately i remembered correctly. `Transfuse` and `Parceler` for instance adds an own version of `@Generated`, because that annotation is not available in the android.jar.

chpasha commented:
I'm not sure I get it :) . This annotation is not included in apk because of it's retention policy. So why should we care about if it is present in android.jar? Like I said, for the sake of experiment I added this annotation to the generation process and it doesn't cause any issues at all.

WonderCsabo commented:
If the annotation is not in the jar, the compilation will fail. This is the case for Eclipse users for instance.

chpasha commented:
I'm not sure I understand what it has to do with eclipse :) . I must be missing something, but the annotation is present in JDK6 and later, so it is available during compilation. But don't get me wrong, I'm just curious, I don't insist on introducing or implementing anything as long as I can maintain my own fork :)

P. S. actually it was added in dagger2 https://github.com/square/dagger/issues/265  as well ;)

WonderCsabo commented:
@chpasha it is indeed JDK6, but that is not in the classpath in case when building an Android project in Eclipse. Only the android.jar is on the classpath.

chpasha commented:
Ok, then maybe adding an annotation processing option that disables @Generated makes sense? Just for the sake of completeness, since this annotation has not much real value anyway.

WonderCsabo commented:
Yeah, that makes sense, and as i said `Lombok` does the same. However we have to disable it default, to not break silently compilation for Eclipse users. Which means AS users will also not find the option, and never generate these annotations. :(
